### ISSUE TITLE
Blank Node Scope

### ISSUE THREAD

**Issue by gkellogg** [json-ld/ndjson-ld#4]
On the call, I think we discussed that blank nodes would be scoped to each individual record, however this would make using this format in a manner similar to N-Quads unable to represent shared blank nodes.

Most serializations scope blank node identifiers to the file. In this reckoning:

```
{"@id": "foo", "prop": {"@id": "_:bar"}}
{"@id": "_:bar", "prop": {"@id": "baz"}}
```

could be semantically equivalent to: 

```
{"@id": "foo", "prop": {"@id": "_:bar", "prop": {"@id": "baz"}}}
```

Otherwise, there would be no way to state this using independent records.

**Proposal: blank node labels are shared across all documents within a stream.**

**Comment by gkellogg** [comment]
<details>
<summary>
This issue was discussed in the <a href="https://json-ld.org/minutes/2022-10-12/#18">2022-10-12 meeting</a>
</summary>
<h2 onmouseout="$('#link-topic-3.1').hide()" onmouseover="$('#link-topic-3.1').show()" id="topic-3.1" class="topic">
Subtopic: Blank Node Scope
<a id="link-topic-3.1" style="display:none;" href="#topic-3.1">✪</a></h2>
<div onmouseout="$('#link-18').hide()" onmouseover="$('#link-18').show()" id="18" class="comment"><span class="name">Gregg Kellogg</span>: <a href="https://github.com/json-ld/ndjson-ld/issues/4">https://github.com/json-ld/ndjson-ld/issues/4</a> <a id="link-18" style="display:none;" href="#18">✪</a></div>
<div onmouseout="$('#link-19').hide()" onmouseover="$('#link-19').show()" id="19" class="comment"><span class="name">Gregg Kellogg</span>:  If every record is a JSON-LD document, it might have its own context. Or it can be provided externally &#8212; as an HTTP header, or an API parameter <a id="link-19" style="display:none;" href="#19">✪</a></div>
<div class="comment-continuation">... Now, regarding the blank nodes scope. We&apos;ve discussed it for YAML-LD before.</div>
<div onmouseout="$('#link-20').hide()" onmouseover="$('#link-20').show()" id="20" class="comment"><span class="name">Gregg Kellogg</span>:  Tag definitions and blank node names might be independent between documents. For instance, if we&apos;re collecting random documents we might have unexpected consequences if the labels or tags overlap. In the streaming applications though, we might want to share labels (for computing differences, etc) but we won&apos;t be able to. <a id="link-20" style="display:none;" href="#20">✪</a></div>
<div onmouseout="$('#link-21').hide()" onmouseover="$('#link-21').show()" id="21" class="comment"><span class="name">Gregg Kellogg</span>:  Next steps on NDJSON-LD &#8212; Nicholas? <a id="link-21" style="display:none;" href="#21">✪</a></div>
<div onmouseout="$('#link-22').hide()" onmouseover="$('#link-22').show()" id="22" class="comment"><span class="name">Niklas Lindstr&#246;m</span>:  I do not lack anything other than time. I should write the simplest thing imaginable first, probably... Blank nodes: I do not have a strong opinion on this, we&apos;re not using blank node identifiers. JSON-LD document can be an RDF dataset and that is an interesting complication. So every line can represent multiple datasets/graphs. <a id="link-22" style="display:none;" href="#22">✪</a></div>
<div onmouseout="$('#link-23').hide()" onmouseover="$('#link-23').show()" id="23" class="comment"><span class="name">Niklas Lindstr&#246;m</span>:  This does not influence blank nodes question though. <a id="link-23" style="display:none;" href="#23">✪</a></div>
<div onmouseout="$('#link-24').hide()" onmouseover="$('#link-24').show()" id="24" class="comment"><span class="name">Niklas Lindstr&#246;m</span>:  In TriG documents, afaik blank node ids are shared throughout the whole document. <a id="link-24" style="display:none;" href="#24">✪</a></div>
<div onmouseout="$('#link-25').hide()" onmouseover="$('#link-25').show()" id="25" class="comment"><span class="name">Gregg Kellogg</span>:  Different use cases might drive conflicting requirements. <a id="link-25" style="display:none;" href="#25">✪</a></div>
<div onmouseout="$('#link-26').hide()" onmouseover="$('#link-26').show()" id="26" class="comment"><span class="name">Leonard Rosenthol</span>:  There is a difference whether we address a homogenuous case (all documents share a context or grammar) vs the case where documents are heteregenuous (potentially unrelated to each other) - but they happen to share one data stream. Do we want to solve both cases? What are we gaining or losing by doing so? <a id="link-26" style="display:none;" href="#26">✪</a></div>
<div onmouseout="$('#link-27').hide()" onmouseover="$('#link-27').show()" id="27" class="comment"><span class="name">Gregg Kellogg</span>:  NDJSON-LD should be an extension of JSON-LD API because YAML-LD calls upon it. Algorithms operate to transform &#8594; JSON-LD internal representation. Consequently, we rely upon that. Regarding the purpose of all this: if we pre-suppose an API and one of this API entry points relates to RDF transformation then doing something multi-dataset becomes really challenging. <a id="link-27" style="display:none;" href="#27">✪</a></div>
<div onmouseout="$('#link-28').hide()" onmouseover="$('#link-28').show()" id="28" class="comment"><span class="name">Gregg Kellogg</span>:  Especially when we think we&apos;re collecting unrelated and maybe conflicting records. It would simplify the problem if all the records relate to a single dataset. Unless we have compelling use cases which suggest otherwise. <a id="link-28" style="display:none;" href="#28">✪</a></div>
<div onmouseout="$('#link-29').hide()" onmouseover="$('#link-29').show()" id="29" class="comment"><span class="name">Gregg Kellogg</span>:  We might introduce a &quot;meta&quot; record to specify meta parameters. Like Turtle with their `&quot;@prefix&quot;` parameters in the header. <a id="link-29" style="display:none;" href="#29">✪</a></div>
<div onmouseout="$('#link-30').hide()" onmouseover="$('#link-30').show()" id="30" class="information"><a href="https://github.com/prefix">https://github.com/prefix</a> -&gt; `@prefix` <a id="link-30" style="display:none;" href="#30">✪</a></div>
<div onmouseout="$('#link-31').hide()" onmouseover="$('#link-31').show()" id="31" class="comment"><span class="name">Niklas Lindstr&#246;m</span>: +1 For &quot;meta-records&quot;; could e.g. set the context initially (corollary: first row of a TSV as columns) <a id="link-31" style="display:none;" href="#31">✪</a></div>
<div onmouseout="$('#link-32').hide()" onmouseover="$('#link-32').show()" id="32" class="comment"><span class="name">Niklas Lindstr&#246;m</span>:  Going to continue the work on NDJSON-LD. <a id="link-32" style="display:none;" href="#32">✪</a></div>
</details>

**Comment by gkellogg** [comment]
In the meeting, we discussed the possibility of some kind of meta-record that could better control blank-node scope. This could be some kind of record with a directive that resets the blank node namespace between records, or records that form a header to the stream document itself that control such behavior for all records.

```
%meta bnode-space disjoint
{"@id": "_:b0", "p1": "v1"}
{"@id": "_:b0", "p2": "v2"}
```

In which the two records do not share the same blank node, even though the use the same identifier, or:

```
%meta bnode-space shared
{"@id": "_:b0", "p1": "v1"}
{"@id": "_:b0", "p2": "v2"}
```

In which both records share the same blank node. Perhaps something like `%meta bnode-space reset` might be used between records to sever the shared space.

For JSON-LD, this would only really come into play when generating RDF, or flattening when blank nodes are renamed, and would be difficult to communicate through the API.

Similar thoughts about if each document describes its own disjoint dataset, or they are all considered sub-sets of a single dataset.

Easiest course would be to say that the stream format is intended for documents that all relate to the same dataset and blank node identifiers are shared across all documents within a stream.

**Comment by TallTed** [comment]
@gkellogg — I *think* you intended for `%meta bnode-space disjoint` to mean "the two records use the same label for _different_ blank nodes, while `%meta bnode-space shared` means "the two records use the same label for _the same_ blank node"?  If not, I'm entirely confused by what the two codeblocks are meant to mean.

**Comment by gkellogg** [comment]
Yes, something got lost in the original, I’ve updated the comment.

**Comment by anatoly-scherbakov** [comment]
This format would require a separate definition for the meta record grammar.

Can this be a JSON record? Even more, can we have a JSON-LD context to define its meaning?

```
{"@context": "https://ndjson-ld.org/meta", "blank-nodes-scope": "global"}
```
